src/js/controller.js

```diff
@@ -0,0 +1,185 @@
+import { validateTodo } from "./filterValidationOnTodo.js";
+import { todoActions, urgency, category } from "./consts.js";
+
+export class Controller {
+  constructor(view, model) {
+    this.model = model;
+    this.view = view;
+
+    this.view.updateHeaderDate();
+    this.view.bindUndo(this.handleUndo);
+    this.view.bindRedo(this.handleRedo);
+    this.view.bindAddTodo(this.handleAddTodo);
+    this.view.bindFilterUpdate(this.handleFilterUpdate);
+    this.view.bindCheckBoxUpdate(this.handleCheckBoxUpdate);
+    this.view.bindSearchBoxUpdate(this.handleSearchBoxUpdate);
+    this.view.bindClearSearchBtn();
+    this.view.bindToggleCompletionOfSelection(this.toggleCompletionOfSelection);
+    this.view.bindClearSelection(this.clearSelection);
+    this.view.bindDeleteSelectedTodos(this.deleteSelectedTodos);
+
+    this.model.bindStateChangeHandler(this.render);
+
+    this.callbacksForTodo = {
+      handleTodoToggle: this.handleTodoToggle,
+      handleTogglingFromSelected: this.handleTogglingFromSelected,
+      handleDeleteTodo: this.handleDeleteTodo,
+      handleEditTodo: this.handleEditTodo,
+    };
+
+    this.render(
+      this.model.allTodos,
+      this.model.currentlySelected,
+      this.model.filterData
+    );
+  }
+
+  render = (todos, currentlySelectedIds, filterData) => {
+    const filteredTodos = todos.filter((todo) =>
+      validateTodo(todo, filterData)
+    );
+    // console.log(filteredTodos);
+    this.view.analyticsUpdater.resetCounts();
+    this.view.displayTodos(
+      filteredTodos,
+      currentlySelectedIds,
+      this.callbacksForTodo
+    );
+    filteredTodos.forEach((todo) =>
+      this.view.analyticsUpdater.updateCounts(todo)
+    );
+    this.view.analyticsUpdater.updateAnalyticsOnView();
+  };
+
+  createTodoObject = ({ counter, title, urgency, category }) => {
+    if (!urgency) {
+      urgency = 0;
+    }
+    if (!category) {
+      category = 0;
+    }
+
+    return {
+      ID: counter,
+      date: new Date().toLocaleString(),
+      title: title,
+      urgency: urgency,
+      category: category,
+      completed: false,
+    };
+  };
+
+  handleAddTodo = (title, urgency, category) => {
+    const todo = this.createTodoObject({
+      counter: this.model.getCounter(),
+      title,
+      urgency,
+      category,
+    });
+    this.model.addTodo(todo);
+  };
+
+  handleFilterUpdate = (urgencyOrCategory, level) => {
+    let index = 0;
+    if (level === urgency.MEDIUM || level === category.ACADEMIC) {
+      index = 1;
+    }
+    if (level == urgency.HIGH || level === category.SOCIAL) {
+      index = 2;
+    }
+    this.model.updateFilter(urgencyOrCategory, index);
```

Comment:
what is `index` for ? index=1 and index=2 doesn't tell me what their purpose is.